File: library/src/main/java/com/afollestad/aesthetic/views/AestheticNavigationView.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.StateListDrawable;
import android.support.annotation.RestrictTo;
import android.support.design.widget.NavigationView;
import android.support.v4.content.ContextCompat;
import android.support.v4.util.Pair;
import android.util.AttributeSet;

import com.afollestad.aesthetic.Aesthetic;
import com.afollestad.aesthetic.NavigationViewMode;
import com.afollestad.aesthetic.R;
import com.afollestad.aesthetic.Util;

import rx.Observable;
import rx.Subscription;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;

/** @author Aidan Follestad (afollestad) */
@RestrictTo(LIBRARY_GROUP)
@SuppressWarnings("RestrictedApi")
public class AestheticNavigationView extends NavigationView {

  private Subscription modeSubscription;
  private Subscription colorSubscription;

  public AestheticNavigationView(Context context) {
    super(context);
  }

  public AestheticNavigationView(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticNavigationView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  private void invalidateColors(Pair<Integer, Boolean> pair) {
    int selectedColor = pair.first;
    boolean isDark = pair.second;
    int baseColor = isDark ? Color.WHITE : Color.BLACK;
    int unselectedIconColor = Util.adjustAlpha(baseColor, .54f);
    int unselectedTextColor = Util.adjustAlpha(baseColor, .87f);
    int selectedItemBgColor =
        ContextCompat.getColor(
            getContext(),
            isDark
                ? R.color.ate_navigation_drawer_selected_dark
                : R.color.ate_navigation_drawer_selected_light);

    final ColorStateList iconSl =
        new ColorStateList(
            new int[][] {
              new int[] {-android.R.attr.state_checked}, new int[] {android.R.attr.state_checked}
            },
            new int[] {unselectedIconColor, selectedColor});
    final ColorStateList textSl =
        new ColorStateList(
            new int[][] {
              new int[] {-android.R.attr.state_checked}, new int[] {android.R.attr.state_checked}
            },
            new int[] {unselectedTextColor, selectedColor});
    setItemTextColor(textSl);
    setItemIconTintList(iconSl);

    StateListDrawable bgDrawable = new StateListDrawable();
    bgDrawable.addState(
        new int[] {android.R.attr.state_checked}, new ColorDrawable(selectedItemBgColor));
    setItemBackground(bgDrawable);
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    modeSubscription =
        Aesthetic.get()
            .navViewMode()
            .compose(distinctToMainThread())
            .subscribe(
                mode -> {
                  switch (mode) {
                    case NavigationViewMode.SELECTED_PRIMARY:
                      colorSubscription =
                          Observable.combineLatest(
                                  Aesthetic.get().primaryColor(),
                                  Aesthetic.get().isDark(),
                                  Pair::create)
                              .compose(distinctToMainThread())
                              .subscribe(this::invalidateColors, onErrorLogAndRethrow());
                      break;
                    case NavigationViewMode.SELECTED_ACCENT:
                      colorSubscription =
                          Observable.combineLatest(
                                  Aesthetic.get().accentColor(),
                                  Aesthetic.get().isDark(),
                                  Pair::create)
                              .compose(distinctToMainThread())
                              .subscribe(this::invalidateColors, onErrorLogAndRethrow());
                      break;
                    default:
                      throw new IllegalStateException("Unknown nav view mode: " + mode);
                  }
                },
                onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    if (modeSubscription != null) {
      modeSubscription.unsubscribe();
    }
    if (colorSubscription != null) {
      colorSubscription.unsubscribe();
    }
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticNavigationView.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.aesthetic.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.StateListDrawable
import android.util.AttributeSet
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.ColorIsDarkState
import com.afollestad.aesthetic.NavigationViewMode.NONE
import com.afollestad.aesthetic.NavigationViewMode.SELECTED_ACCENT
import com.afollestad.aesthetic.NavigationViewMode.SELECTED_PRIMARY
import com.afollestad.aesthetic.R
import com.afollestad.aesthetic.utils.adjustAlpha
import com.afollestad.aesthetic.utils.color
import com.afollestad.aesthetic.utils.combine
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.subscribeTo
import com.afollestad.aesthetic.utils.unsubscribeOnDetach
import com.google.android.material.navigation.NavigationView
import io.reactivex.Observable.empty
import io.reactivex.disposables.Disposable

/** @author Aidan Follestad (afollestad) */
@SuppressLint("RestrictedApi")
class AestheticNavigationView(
  context: Context,
  attrs: AttributeSet? = null
) : NavigationView(context, attrs) {

  private var colorSubscription: Disposable? = null

  private fun invalidateColors(state: ColorIsDarkState) {
    val selectedColor = state.color
    val isDark = state.isDark
    val baseColor = if (isDark) Color.WHITE else Color.BLACK
    val unselectedIconColor = baseColor.adjustAlpha(.54f)
    val unselectedTextColor = baseColor.adjustAlpha(.87f)

    val selectedItemBgColor = context.color(
        if (isDark) R.color.ate_navigation_drawer_selected_dark
        else R.color.ate_navigation_drawer_selected_light
    )

    val iconSl = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_checked),
            intArrayOf(android.R.attr.state_checked)
        ),
        intArrayOf(unselectedIconColor, selectedColor)
    )
    val textSl = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_checked),
            intArrayOf(android.R.attr.state_checked)
        ),
        intArrayOf(unselectedTextColor, selectedColor)
    )

    itemTextColor = textSl
    itemIconTintList = iconSl

    val bgDrawable = StateListDrawable()
    bgDrawable.addState(
        intArrayOf(android.R.attr.state_checked),
        ColorDrawable(selectedItemBgColor)
    )
    itemBackground = bgDrawable
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    get().navigationViewMode()
        .distinctToMainThread()
        .flatMap {
          when (it) {
            SELECTED_PRIMARY -> combine(
                get().colorPrimary(),
                get().isDark
            ) { color, isDark -> ColorIsDarkState(color, isDark) }

            SELECTED_ACCENT -> combine(
                get().colorAccent(),
                get().isDark
            ) { color, isDark -> ColorIsDarkState(color, isDark) }

            NONE -> empty()
          }
        }
        .distinctUntilChanged()
        .subscribeTo(::invalidateColors)
        .unsubscribeOnDetach(this)
  }

  override fun onDetachedFromWindow() {
    colorSubscription?.dispose()
    super.onDetachedFromWindow()
  }
}
